apisix/plugins/csrf.lua

```diff
@@ -0,0 +1,161 @@
+            default = 7200
+        },
+        name = {
+            description = "the csrf token name",
+            type = "string",
+            default = "apisix-csrf-token"
+        }
+    },
+    required = {"key"}
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = 2980,
+    name = "csrf",
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+local function gen_sign(random, expires, key)
+    local sha256 = resty_sha256:new()
+
+    local sign = "{expires:" .. expires .. ",random:" .. random .. ",key:" .. key .. "}"
+
+    sha256:update(sign)
+    local digest = sha256:final()
+
+    return str.to_hex(digest)
+end
+
+
+local function gen_csrf_token(conf)
+    local random = math.random()
+    local sign = gen_sign(random, conf.expires, conf.key)
+
+    local token = {
+        random = random,
+        expires = conf.expires,
+        sign = sign,
+    }
+
+    local cookie = ngx_encode_base64(core.json.encode(token))
+    return cookie
+end
+
+
+local function check_csrf_token(conf, ctx, token)
+    local token_str = ngx_decode_base64(token)
+    if token_str == nil then
+        core.log.error("csrf token base64 decode error")
+        return false
+    end
+
+    local token_table, err = core.json.decode(token_str)
+    if err then
+        core.log.error("decode token error: ", err)
+        return false
+    end
+
+    local random = token_table["random"]
+    if not random then
+        core.log.error("no random in token")
+        return false
+    end
+
+    local expires = token_table["expires"]
+    if not expires then
+        core.log.error("no expires in token")
+        return false
+    end
+
+    local sign = gen_sign(random, expires, conf.key)
+    if token_table["sign"] ~= sign then
+        core.log.error("Invalid signatures")
+        return false
+    end
+
+    return true
+end
+
+
+function _M.access(conf, ctx)
+    local safe_methods = {"GET", "HEAD", "OPTIONS"}
+    local method = core.request.get_method(ctx)
+    if core.table.array_find(safe_methods, method) then
+        return
+    end
+
+    local header_token = core.request.header(ctx, conf.name)
+    if not header_token or header_token == "" then
+        return 401, {error_msg = "no csrf token in headers"}
+    end
+
+    local cookie_token = ctx.var["cookie_" .. conf.name]
+    if not cookie_token then
+        return 401, {error_msg = "no csrf cookie"}
+    end
+
+    if header_token ~= cookie_token then
+        return 401, {error_msg = "csrf token mismatch"}
+    end
+
+    local result = check_csrf_token(conf, ctx, cookie_token)
+    if not result then
+        return 401, {error_msg = "Failed to verify the csrf token signature"}
+    end
+end
+
+
+function _M.header_filter(conf, ctx)
+    local csrf_token = gen_csrf_token(conf)
+    local cookie = conf.name .. "=" .. csrf_token .. ";path=/;SameSite=Lax;Expires="
+                   .. cookie_time(ngx_time() + conf.expires)
```

Comment:
Instead of `Expires` I think we can use `Max-Age` attribute here. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie

Since, the operation is performed for each request, at the long rung it will save some computation.